### DISCUSSION TITLE
why can I not prove this (trivial) lemma?

### DISCUSSION THREAD

**Discussion by Meijster** [dafny-lang/dafny#4296] [Q&A]
Hi everyone,

I tried a proof and Dafny is not able to prove it. I stripped it down to the following:
```
lemma trivial(n: nat)
  ensures true
{
  var a := n*(n+1)/2;
  assert a == n*(n+1)/2;
  assert 2*a == n*(n+1);
}
```
The first assert is (of course) accepted, but the second fails (after quite a while).
If I ask for a counter example (press F7 in visual studio code), then it comes up with n: int = 520 a: int = 135460.

Why does this verification fail, and how is this a counterexample?

**Comment by zafer-esen** [comment]
This seems related to #488.  the multiplication `n*(n+1)` seems to produce an incorrect value. For instance:
```
var a := (n*(n+1))/2;
var rhs := n*(n+1);
var lhs := 2*a;
assert 2*a == n*(n+1);
```

Counterexample `n = 515`, `rhs = 265741` (off by one) and `lhs = 265740`.

**Comment by MikaelMayer** [comment]
Since Dafny is a proof engine and not a model checker, the counter examples are rather a model we extract from Z3 after Z3 told us "I couldn't disprove that this counterexample is not valid", so unfortunately they cannot be relied upon and are not a proof of failure like in model checkers. They can give hints at the best. We recentely re-labelled them as "experimental" to lower expectations, until we have a better name or interface.

Note that you'd first need to prove that n*(n+1) is even so that the division is an integer one.

Try adding this and prove it:

```
assert (n * (n+1)) % 2 == 0;
```

**Reply by Meijster** [reply]
Thanks for your answer. Still I am struggling. I tried your suggestion but fail in doing so.
Just adding your assert does not help. So, I tried:

```
lemma trivial(n: nat)
  ensures true
{
  if (n%2 == 0) {
    assert n%2 == 0;   // succeeds
    assert (n*(n+1))%2 == 0;  // fails
  } else {
    assert (n+1)%2 == 0; // succeeds
    assert (n*(n+1))%2 == 0;  // fails
  }
  assert (n * (n+1)) % 2 == 0;
}
```

I think the cause is that Dafny does not know (cannot proof) that if a x is even then also k*x is even.
But I do not see how to make it understand that.

Can you tell/explain me how to do it right?

**Reply by zafer-esen** [reply]
Someone might be able to provide a better answer, but adding in the following two lemmas helped Dafny prove this.

```
// call consecMultAlwaysEven(n) from the trivial lemma.

// Show that multiplying any even number with
// a number will yield an even number.
// This proof uses induction.
lemma evenMult (n : nat, k : nat)
  requires n % 2 == 0
  ensures (n * k) % 2 == 0
{
  if (k == 0) {
    // Dafny automatically proves the base case.
  } 
  else if (k > 0) {
    evenMult(n, k-1);
    assert((n * (k-1)) % 2 == 0);
  }
}

// Show that the multiplication of two consecutive
// numbers are always even.
lemma consecMultAlwaysEven (n : nat)
  ensures (n * (n+1)) % 2 == 0
{
  if (n % 2 == 0) {
    evenMult(n, n+1);
  } else {
    evenMult(n+1, n);
  }
}
```
Part of the proof uses induction. [Here is some reading](https://dafny.org/dafny/OnlineTutorial/Lemmas) about lemmas and induction in Dafny that might be useful. There is also [a recent blog post](https://dafny.org/blog/2023/07/14/types-and-programming-languages/) from @MikaelMayer which has lots of great examples that make use of induction.

**Comment by Meijster** [comment]
OK, thank you. That helped!
